src/main/kotlin/app/batch/Validator.kt

```diff
@@ -0,0 +1,87 @@
+package app.batch
+
+import app.domain.DecryptedRecord
+import app.domain.SourceRecord
+import app.exceptions.BadDecryptedDataException
+import com.google.gson.Gson
+import com.google.gson.JsonObject
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import java.text.SimpleDateFormat
+import java.util.*
+
+@Component
+class Validator {
+
+    fun skipBadDecryptedRecords(item: SourceRecord, decrypted: String): DecryptedRecord? {
+        val hbaseRowKey = Arrays.copyOfRange(item.hbaseRowId, 4, item.hbaseRowId.size)
+        val hbaseRowId = String(hbaseRowKey)
+        val db = item.db
+        val collection = item.collection
+        try {
+            val jsonObject = parseDecrypted(hbaseRowId, decrypted)
+            if (null != jsonObject) {
+                retrieveId(hbaseRowId, jsonObject)
+                val lastUpdatedTimestamp = retrievelastUpdatedTimestamp(hbaseRowId, jsonObject)
+                lastUpdatedTimestamp?.let { validate(hbaseRowId, lastUpdatedTimestamp) }
+                jsonObject.addProperty("timestamp", item.hbaseTimestamp)
+                return DecryptedRecord(jsonObject, db, collection)
+            }
+        } catch (e: Exception) {
+            val ex = BadDecryptedDataException(hbaseRowId, db, collection, e.message!!)
+            logger.error(ex.message)
+            throw ex
+        }
+        return null
+    }
+
+    fun parseDecrypted(hbaseRowId: String, decrypted: String): JsonObject? {
+        try {
+            val jsonObject = Gson().fromJson(decrypted, JsonObject::class.java)
+            return jsonObject
+        } catch (e: Exception) {
+            val parsingException = "Exception occurred while parsing decrypted db object"
+            throw Exception(parsingException)
+        }
+        return null
+    }
+
+    fun retrieveId(hbaseRowId: String, jsonObject: JsonObject): JsonObject? {
+        val id = jsonObject.getAsJsonObject("_id")
+        if (null == id) {
+            val idNotFound = "id not found in the decrypted db object"
+            throw Exception(idNotFound)
+        }
+        return id
+    }
+
+    fun retrievelastUpdatedTimestamp(hbaseRowId: String, jsonObject: JsonObject): JsonObject? {
+        val lastUpdatedTimestamp = jsonObject.getAsJsonObject("_lastModifiedDateTime")
+        if (null == lastUpdatedTimestamp) {
+            val _lastModifiedDateTimeNotFound = "_lastModifiedDateTime not found in the decrypted db object"
+            throw Exception(_lastModifiedDateTimeNotFound)
+        }
+        return lastUpdatedTimestamp
+    }
+
+    fun validate(hbaseRowId: String, lastUpdatedTimestamp: JsonObject): Long? {
```

Comment:
This function name is a bit misleading - it returns the timestamp as a long yet it is called 'validate'

Reply:
renamed for better  readability